### PR TITLE
Add documentation for plugin development

### PR DESCRIPTION
**Pull request by bpjordan**

### CODE REVIEW

**Review by carterray31001 — COMMENTED** [review]
Few questions and comments:
 - can you add some information about what format the request is in when it's received by say an interceptor or listener module
    - like do I have to decode it myself in some instances if writing certain modules?
 - also, I want to talk about the decision behind having the listener respond with the plug-in and having the other modules returning codes as to how it was resolved
 - what should frontend do with bad API responses from plugins
 - missing section about plug-in logo implementation

**Comment by bpjordan** [comment]
> * can you add some information about what format the request is in when it's received by say an interceptor or listener module
>   * like do I have to decode it myself in some instances if writing certain modules?

Could you clarify what you're wanting to know? There's a section talking about formatting of DNS messages in the section formatting of provided types, including DNS messages.

> * also, I want to talk about the decision behind having the listener respond with the response by being polled periodically 

I'm open to discussion on this but I feel pretty strongly that the async (i.e. polling) method is the way to go. From an architectural standpoint the current method of decoding requests from a socket managed by the server is heavily limiting for plugin authors and would require a lot of support in order to support use cases other than the most basic method we're doing now (we don't even currently support TCP because I haven't had time to implement it yet).

> and having the other modules returning codes as to how it was resolved

I have much fewer strong feelings about this one, but I think it does provide some nice ergonomics & logging capabilities.

> * what should frontend do with bad API responses from plugins

I was envisioning a red banner that just copies the error message provided by the plugin

> * missing section about plug-in logo implementation

Fix incoming

**Comment by carterray31001** [comment]
> > * can you add some information about what format the request is in when it's received by say an interceptor or listener module
> 
> >   * like do I have to decode it myself in some instances if writing certain modules?
> 
> 
> 
> Could you clarify what you're wanting to know? There's a section talking about formatting of DNS messages in the section formatting of provided types, including DNS messages. Do I need to clarify something in there?
> 
I was mainly just asking about the response format through the plug-in modules. I assume listener has to decode and resolver has to encode?

> 
> > * also, I want to talk about the decision behind having the listener respond with the response by being polled periodically 
> 
> 
> 
> I'm open to discussion on this but I feel pretty strongly that the async (i.e. polling) method is the way to go. From an architectural standpoint the current method of decoding requests from a socket managed by the server is heavily limiting for plugin authors and would require a lot of server support in order to support use cases other than the most basic method we're doing now.
> 
> 
> Right now, we are assuming that all DNS traffic comes over UDP port 53, which prevents us from supporting DNS over HTTP, DNS over TLS, larger standard DNS requests (which are sent over TCP), etc. Under the current method, adding support for these server-side would mean adding infrastructure to bind user-defined ports to a specific listener, which would be cumbersome both from an architectural standpoint and from a UX standpoint.
> 
> 
> As an example, we don't even currently support TCP because I haven't had time to implement it yet. Using the async method, it would be trivial for a plugin to listen on both  TCP and UDP, by binding its own TCP and UDP listeners and polling them each time the plugin itself is polled.
> 
> 
> 
> Even if we did this, some standard DNS methods still wouldn't be supported because they would require some handshake (such as the TLS handshake for DNS over HTTPS). With a server-managed socket, the plugin would have no way of specifying that something like this is required.
> 
> 

I definitely don't disagree that async is the way to go, the main question I had with it is why does the listener respond and not whichever plug-in resolves it originally? Is this because of the polling?

> 
> > and having the other modules returning codes as to how it was resolved
> 
> 
> 
> I assume here you're talking about allowing API implementation functions to determine their own API response code?
> 
> 
> 
> I have much fewer strong feelings about this one, but I think it does provide some nice ergonomics & logging capabilities. From a frontend standpoint, I think it is a simple question of treating something as an error if it is an error code and treating something as successful if it is a success code.
> 
> 

This one I was more referring back to how the listener resolves it with different codes 0-2 about what plug-in module resolved it. I was wanting to hear your opinions on the difference between just having each plug-in respond to the request accordingly vs having the listener respond and why the code we have will work as we are intending.

> 
> > * what should frontend do with bad API responses from plugins
> 
> 
> 
> I was envisioning a red banner that just copies the error message provided by the plugin
> 
> 
> 
> > * missing section about plug-in logo implementation
> 
> 
> 
> Fix incoming
> 
> 
> 

Basically I'm hoping to chat about some of this today and get my brain on the same page as yours so I can understand more of what the plug-in creation process would look like.

**Comment by bpjordan** [comment]
Looks like the main unresolved question is "why are listeners responsible for responding to a request rather than other modules which actually generate the response?"

Since the purpose of having listeners is to allow plugins to determine what channels can be used to make DNS requests, it necessarily follows that the listener that decoded the request would also need  to encode the request, since it knows how to do so.

For example, if a request comes over DNS over HTTPS, then an interceptor which encodes a response to that request wouldn't know how to respond to it. Thus we send the response back to the listener that initially received the request, since it does know how to send that request back to the requester.

**Comment by carterray31001** [comment]
I see! So, when having multiple listeners the response is sent back to the original listener for reencoding

**Review by tchuhn — APPROVED** [review]

**Review by carterray31001 — APPROVED** [review]
Beautiful
